numpy/core/src/umath/ufunc_object.c

```diff
@@ -924,30 +935,67 @@ convert_ufunc_arguments(PyUFuncObject *ufunc,
     PyObject *obj;
 
     /* Convert and fill in input arguments */
+    int all_scalar = 1;
+    int any_scalar = 0;
+    int all_legacy = 1;
     for (int i = 0; i < nin; i++) {
         obj = PyTuple_GET_ITEM(full_args.in, i);
 
         if (PyArray_Check(obj)) {
-            PyArrayObject *obj_a = (PyArrayObject *)obj;
-            out_op[i] = (PyArrayObject *)PyArray_FromArray(obj_a, NULL, 0);
+            out_op[i] = (PyArrayObject *)obj;
+            Py_INCREF(out_op[i]);
+        }
+        else {
+            /* Convert the input to an array and check for special cases */
+            out_op[i] = (PyArrayObject *)PyArray_FromAny(obj, NULL, 0, 0, 0, NULL);
+            if (out_op[i] == NULL) {
+                goto fail;
+            }
+        }
+        out_borrowed_op_DTypes[i] = NPY_DTYPE(PyArray_DESCR(out_op[i]));
+        if (!out_borrowed_op_DTypes[i]->legacy) {
```

Comment:
I guess the fact that this is not covered answers my question elsewhere, that there are no tests yet for the new path. I wonder if there shouldn't be at least some simple ones!?

Reply:
I am operating under the thought of getting NumPy ready first and writing tests when this becomes closer to public API, or as we might use it internally even.
That could be as part of an "example"  DType.  For me that just seemed easier to keep things moving, even if it is bound to lead to some fixups being necessary later.

But if it seems better or helps with understanding the code, I can try to start adding them here I guess.

That said, I do not think I can cover *all* branches yet and leaving them out is worse IMO.  But, I could cover some of using the private API.  (For one, I omitted promotion, which would be good to have for tests.)

Reply:
I think it would be very useful to have tests for things that already work, if only to help review! For my own code, I find that the tests sometimes become better documentation than the docstrings... Plus, they invariably smoke out errors. And, as you note, they end up being useful as examples in documentation. So, yes, if you can, please do add some tests!

Reply:
Well, working on it (just pushed, will probably fail the tests).  I need casting etc. as well (which somewhat would work without ufuncs)...  Unfortunately it is 660 lines of C code, although mostly straight forward probably.

The idea was to use a "scaled" float/double (not really useful, but a bit like a unit).